File: android/java/org/chromium/chrome/browser/crypto_wallet/KeyringControllerFactory.java
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.crypto_wallet;

import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.brave_wallet.mojom.KeyringController;
import org.chromium.mojo.bindings.ConnectionErrorHandler;
import org.chromium.mojo.bindings.Interface;
import org.chromium.mojo.bindings.Interface.Proxy.Handler;
import org.chromium.mojo.system.MessagePipeHandle;
import org.chromium.mojo.system.MojoException;
import org.chromium.mojo.system.impl.CoreImpl;

@JNINamespace("chrome::android")
public class KeyringControllerFactory {
    private static final Object lock = new Object();
    private static KeyringControllerFactory instance;

    public static KeyringControllerFactory getInstance() {
        synchronized (lock) {
            if (instance == null) {
                instance = new KeyringControllerFactory();
            }
        }
        return instance;
    }

    private KeyringControllerFactory() {}

    public KeyringController getKeyringController(ConnectionErrorHandler connectionErrorHandler) {
        int nativeHandle = KeyringControllerFactoryJni.get().getInterfaceToKeyringController();
        MessagePipeHandle handle = wrapNativeHandle(nativeHandle);
        KeyringController keyringController = KeyringController.MANAGER.attachProxy(handle, 0);
        Handler handler = ((Interface.Proxy) keyringController).getProxyHandler();
        handler.setErrorHandler(connectionErrorHandler);

        return keyringController;
    }

    private MessagePipeHandle wrapNativeHandle(int nativeHandle) {
        return CoreImpl.getInstance().acquireNativeHandle(nativeHandle).toMessagePipeHandle();
    }

    @NativeMethods
    interface Natives {
        int getInterfaceToKeyringController();
    }
}

File: browser/brave_wallet/keyring_controller_factory.cc
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "brave/browser/brave_wallet/keyring_controller_factory.h"

#include "brave/browser/brave_wallet/brave_wallet_context_utils.h"
#include "brave/components/brave_wallet/browser/keyring_controller.h"
#include "chrome/browser/profiles/incognito_helpers.h"
#include "components/keyed_service/content/browser_context_dependency_manager.h"
#include "components/user_prefs/user_prefs.h"

namespace brave_wallet {

// static
KeyringControllerFactory* KeyringControllerFactory::GetInstance() {
  return base::Singleton<KeyringControllerFactory>::get();
}

// static
mojo::PendingRemote<mojom::KeyringController>
KeyringControllerFactory::GetForContext(content::BrowserContext* context) {
  if (!IsAllowedForContext(context))
    return mojo::PendingRemote<mojom::KeyringController>();

  return static_cast<KeyringController*>(
             GetInstance()->GetServiceForBrowserContext(context, true))
      ->MakeRemote();
}

// static
KeyringController* KeyringControllerFactory::GetControllerForContext(
    content::BrowserContext* context) {
  if (!IsAllowedForContext(context)) {
    return nullptr;
  }
  return static_cast<KeyringController*>(
      GetInstance()->GetServiceForBrowserContext(context, true));
}

KeyringControllerFactory::KeyringControllerFactory()
    : BrowserContextKeyedServiceFactory(
          "KeyringController",
          BrowserContextDependencyManager::GetInstance()) {}

KeyringControllerFactory::~KeyringControllerFactory() {}

KeyedService* KeyringControllerFactory::BuildServiceInstanceFor(
    content::BrowserContext* context) const {
  return new KeyringController(user_prefs::UserPrefs::Get(context));
}

content::BrowserContext* KeyringControllerFactory::GetBrowserContextToUse(
    content::BrowserContext* context) const {
  return chrome::GetBrowserContextRedirectedInIncognito(context);
}

}  // namespace brave_wallet
